ruma-identifiers/src/mxc.rs

```diff
@@ -0,0 +1,119 @@
+//! Matrix-spec compliant mxc:// urls.
+
+use std::{
+    convert::TryFrom,
+    fmt::{self, Display},
+    str::FromStr,
+};
+
+use ruma_identifiers_validation::mxc::validate;
+
+use crate::{Error, MediaId, MediaIdBox, ServerName, ServerNameBox};
+
+/// Matrix-spec compliant mxc:// urls.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Mxc {
+    server_name: ServerNameBox,
+    media_id: MediaIdBox,
```

Comment:
Since string conversions are much more common than accessing individual parts of identifiers we've generally been storing the entire string as one thing. Also I'm planning to soon remove all additional metadata entirely such that the ID types are ZSTs (zero-sized types), but you don't need to do the same in this PR.

Reply:
I think in the case of the `mxc` we receive the a the url, but use pretty much always the parts. Look at the media requests they all require the server name and the media id as separate parameter.

Reply:
Hrm, this would be the only type that can't become a ZST then. I'll have to think about it...

Reply:
I guess modifying the media requests to accept the `mxc://`  directly isn't really a viable option, right?

Reply:
It might be viable, but I think that's tangential to how we store the `mxc` URLs.

Reply:
It would remove the need of accessing the parts of the `mxc://`.